### PR TITLE
add pop sizing

### PR DESCRIPTION
**Pull request by swethalohith2**
## Why
Customers often ask how to size the PoP hardware configuration before installing the PoP

## What
Add pop sizing command

## References
[Story](https://instana.kanbanize.com/ctrl_board/132/cards/159796/details/)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -97,7 +97,13 @@ Upgrade `synctl` to a specified version.
 pip3 install --upgrade synctl==<version>
 ```
 
+# Size the PoP hardware configuration
+synctl can be used to find the size of the PoP hardware configuration.
```
Comment (rui1234) [inline README.md:101]: change "find" to "estimate"

File: README.md [hunk 2]
```diff
@@ -97,7 +97,13 @@ Upgrade `synctl` to a specified version.
...
 
+### synctl pop-sizing Syntax
+```
+synctl --pop-sizing-estimate
+```
```
Comment (rongzhus) [inline README.md:106]: Hi @rui1234 @swethalohith2 I recommend we use a sub command not a single option here, for example `synctl get cost` or use a separate command for cost estimation like `synctl cost estimate --options`, to keep same with other command, and it is easy to extend more options with command.

File: synctl/cli.py [hunk 3]
```diff
@@ -331,6 +332,119 @@ def exit_synctl(self, error_code=-1, message=''):
...
+        pod_estimate = int(user_tests * default_frequency) / int(user_frequency * default_tests)
+        return math.ceil(pod_estimate)
+
+
+    def pop_size_estimate(self):
+        print("Please answer below questions for estimating the self-hosted PoP hardware size\n")
+        try:
+            api_simple = int(self.ask_question("How many API Simple tests do you want to create? (0 if no) "))
+            if api_simple != 0:
+                api_simple_frequency = int(self.ask_question("What is the test frequency for your API Simple tests ? (1-120)  "))
```
Comment (rongzhus) [inline synctl/cli.py:399]: check the value of `api_simple_frequency` here, it should be between 1 and 120, any value other than [1, 120] should output a warning message and ask user to input it again.

File: synctl/cli.py [hunk 4]
```diff
@@ -331,6 +332,119 @@ def exit_synctl(self, error_code=-1, message=''):
...
+                http_pod_count = int(self.size_estimate(api_simple, self.http["frequency"], api_simple_frequency, self.http["testCount"]))
+            else:
+                http_pod_count = 0
+
+            api_script = int(self.ask_question("How many API Script tests do you want to create? (0 if no) "))
+            if api_script != 0:
+                api_script_frequency = int(self.ask_question("What is the test frequency for your API Script tests ? (1-120) "))
```
Comment (rongzhus) [inline synctl/cli.py:406]: same issue verify its issue `api_script_frequency`

File: synctl/cli.py [hunk 5]
```diff
@@ -331,6 +332,119 @@ def exit_synctl(self, error_code=-1, message=''):
...
+        print("Please answer below questions for estimating the self-hosted PoP hardware size\n")
+        try:
+            api_simple = int(self.ask_question("How many API Simple tests do you want to create? (0 if no) "))
+            if api_simple != 0:
+                api_simple_frequency = int(self.ask_question("What is the test frequency for your API Simple tests ? (1-120)  "))
+                http_pod_count = int(self.size_estimate(api_simple, self.http["frequency"], api_simple_frequency, self.http["testCount"]))
+            else:
+                http_pod_count = 0
+
+            api_script = int(self.ask_question("How many API Script tests do you want to create? (0 if no) "))
```
Comment (rongzhus) [inline synctl/cli.py:404]: verify the value of `api_script`, when I test it, I found it can accept a nagative value.

File: synctl/cli.py [hunk 6]
```diff
@@ -331,6 +332,119 @@ def exit_synctl(self, error_code=-1, message=''):
...
+            if api_script != 0:
+                api_script_frequency = int(self.ask_question("What is the test frequency for your API Script tests ? (1-120) "))
+                javascript_pod_count = int(self.size_estimate(api_script, self.javascript["frequency"], api_script_frequency, self.javascript["testCount"]))
+            else:
+                javascript_pod_count = 0
+
+            browser_script = int(self.ask_question("How many Browser tests (Webpage Action, Webage Script and BrowserScript) do you want to create? (0 if no) "))
```
Comment (rongzhus) [inline synctl/cli.py:411]:
```suggestion
            browser_script = int(self.ask_question("How many Browser tests (Webpage Action, Webpage Script and BrowserScript) do you want to create? (0 if no) "))
```
typo

**Comment by rongzhus** [comment]
For final result, we can separate to 4 parts, below is an example, it is more clear:
```
Please answer below questions for estimating the self-hosted PoP hardware size:

How many API Simple tests do you want to create? (0 if no) 100
What is the test frequency for your API Simple tests? (1-120) 1
How many API Script tests do you want to create? (0 if no) 10
What is the test frequency for your API Script tests? (1-120) 5
How many Browser tests (Webpage Action, Webpage Script and BrowserScript) do you want to create? (0 if no) 10
What is the test frequency for Browser tests? (1-120) 5
Do you want to install the Instana-agent to monitor your PoP? (Y/N) Y
How many worker nodes in your kubernetes cluster? 3

Your requirement is:
    API Simple: 1000    Frequency: 1min
    API Script: 50      Frequency: 5min
    Browser Test: 50    Frequency: 10min
    Install Agent: Yes

The estimated sizing is:
    CPU: 15700m
    Mem: 14242MiB
    Disk: 7240GB

Recommend Playback engine pods are:
    http        playback engines: 1 pod
    javascript  playback engines: 1 pod
    browser     playback engines: 2 pod

```

File: synctl/cli.py [hunk 7]
```diff
@@ -331,6 +333,148 @@ def exit_synctl(self, error_code=-1, message=''):
...
+        pod_estimate = int(user_tests * default_frequency) / int(user_frequency * default_tests)
+        return math.ceil(pod_estimate)
+
+    def pop_size_estimate(self):
+        print("Please answer below questions for estimating the self-hosted PoP hardware size\n")
+        try:
+            api_simple = int(self.ask_question("How many API Simple tests do you want to create? (0 if no) "))
+            if api_simple > 0:
+                while True:
+                    api_simple_frequency = int(self.ask_question("What is the test frequency for your API Simple tests ? (1-120)  "))
```
Comment (rongzhus) [inline synctl/cli.py:400]:
```suggestion
                    api_simple_frequency = int(self.ask_question("What is the test frequency for your API Simple tests? (1-120)  "))
```

File: synctl/cli.py [hunk 8]
```diff
@@ -331,6 +333,148 @@ def exit_synctl(self, error_code=-1, message=''):
...
+                        break
+                    else:
+                        print("frequency is not valid, it should be in [1,120]")
+            else:
+                http_pod_count = 0
+
+            api_script = int(self.ask_question("How many API Script tests do you want to create? (0 if no) "))
+            if api_script > 0:
+                while True:
+                    api_script_frequency = int(self.ask_question("What is the test frequency for your API Script tests ? (1-120) "))
```
Comment (rongzhus) [inline synctl/cli.py:412]:
```suggestion
                    api_script_frequency = int(self.ask_question("What is the test frequency for your API Script tests? (1-120) "))
```

File: synctl/cli.py [hunk 9]
```diff
@@ -331,6 +333,148 @@ def exit_synctl(self, error_code=-1, message=''):
...
+                        break
+                    else:
+                        print("frequency is not valid, it should be in [1,120]")
+            else:
+                javascript_pod_count = 0
+
+            browser_script = int(self.ask_question("How many Browser tests (Webpage Action, Webpage Script and BrowserScript) do you want to create? (0 if no) "))
+            if browser_script > 0:
+                while True:
+                    browser_script_frequency = int(self.ask_question("What is the test frequency for Browser tests ? (1-120) "))
```
Comment (rongzhus) [inline synctl/cli.py:424]:
```suggestion
                    browser_script_frequency = int(self.ask_question("What is the test frequency for Browser tests? (1-120) "))
```

**Comment by rongzhus** [comment]
When all test number is 0, there are still an estimated size of resources , is it default value? In such case, the command can exit and show a warning message.
image

File: synctl/cli.py [hunk 10]
```diff
@@ -331,6 +333,148 @@ def exit_synctl(self, error_code=-1, message=''):
...
+                     self.redis["memLimit"] * redis_pod_count
+
+            disk_size = http_pod_count * self.http["imageSize"] + javascript_pod_count * self.javascript["imageSize"] + \
+                        browserscript_pod_count * self.browserscript["imageSize"] + controller_pod_count * self.controller["imageSize"] + \
+                        redis_pod_count * self.redis["imageSize"] + worker_nodes * self.agent["imageSize"] + \
+                        k8ssensor_pod_count * self.k8ssensor["imageSize"]
+
+            max_label_length = max(len(str(api_simple)), len(str(api_script)), len(str(browser_script)), len(str(agent)))
+            print("\nYour requirement is:")
+            print(f"   API Simple: {api_simple:<{max_label_length}}       Frequency: {api_simple_frequency}" if api_simple > 0 else f"   API Simple: {api_simple:<{max_label_length}}")
```
Comment (rongzhus) [inline synctl/cli.py:461]:
```suggestion
            print(f"   API Simple: {api_simple:<{max_label_length}}       Frequency: {api_simple_frequency}min" if api_simple > 0 else f"   API Simple: {api_simple:<{max_label_length}}")
```

File: README.md [hunk 11]
```diff
@@ -97,7 +97,13 @@ Upgrade `synctl` to a specified version.
 pip3 install --upgrade synctl==<version>
 ```
 
+# Size the PoP hardware configuration
+synctl can be used to estimate the size of the PoP hardware configuration.
 
+### synctl pop-sizing Syntax
+```
+synctl --pop-sizing-estimate
```
Comment (rongzhus) [inline README.md:105]:
```suggestion
synctl get pop-size
```

**Comment by swethalohith2** [comment]
> When all test number is 0, there are still an estimated size of resources , is it default value? In such case, the command can exit and show a warning message. image

This is because there is default pods for controller, Redis and k8ssensor.

File: synctl/cli.py [hunk 12]
```diff
@@ -331,6 +333,148 @@ def exit_synctl(self, error_code=-1, message=''):
...
+            "cpuLimit" : 800,
+            "memLimit" : 300,
+            "imageSize" : 400,
+        }
+        self.browserscript = {
+            "testCount" : 5,
+            "frequency": 5,
+            "cpuLimit" : 4000,
+            "memLimit" : 3000,
+            "imageSize" : 1500,
```
Comment (rongzhus) [inline synctl/cli.py:378]: Remove the space before `:` to make format be consistent with other code.

File: synctl/cli.py [hunk 13]
```diff
@@ -331,6 +333,150 @@ def exit_synctl(self, error_code=-1, message=''):
...
+                redis_pod_count = 1
+                k8ssensor_pod_count = 3
+
+                cpu = self.controller["cpuLimit"] * controller_pod_count + self.redis["cpuLimit"] * redis_pod_count + http_pod_count * self.http["cpuLimit"] + \
+                      javascript_pod_count * self.javascript["cpuLimit"] + browserscript_pod_count * self.browserscript["cpuLimit"] + \
+                      worker_nodes * self.agent["cpuLimit"] + self.k8ssensor["cpuLimit"] * k8ssensor_pod_count
+
+                memory = http_pod_count * self.http["memLimit"] + javascript_pod_count * self.javascript["memLimit"] + \
+                         browserscript_pod_count * self.browserscript["memLimit"] + worker_nodes * self.agent["memLimit"] +  \
+                         self.k8ssensor["memLimit"] * k8ssensor_pod_count +  self.controller["memLimit"] * controller_pod_count + \
```
Comment (rongzhus) [inline synctl/cli.py:453]: does sensor memory is added to final result when agent installation is `N`?

File: synctl/cli.py [hunk 14]
```diff
@@ -331,6 +333,150 @@ def exit_synctl(self, error_code=-1, message=''):
...
+                         self.redis["memLimit"] * redis_pod_count
+
+                disk_size = http_pod_count * self.http["imageSize"] + javascript_pod_count * self.javascript["imageSize"] + \
+                            browserscript_pod_count * self.browserscript["imageSize"] + controller_pod_count * self.controller["imageSize"] + \
+                            redis_pod_count * self.redis["imageSize"] + worker_nodes * self.agent["imageSize"] + \
+                            k8ssensor_pod_count * self.k8ssensor["imageSize"]
```
Comment (rongzhus) [inline synctl/cli.py:459]: >k8ssensor_pod_count * self.k8ssensor["imageSize"]

same issue as memory

**Comment by rongzhus** [comment]
> > When all test number is 0, there are still an estimated size of resources , is it default value? In such case, the command can exit and show a warning message. image
> 
> This is because there is default pods for controller, Redis and k8ssensor.

It does not make sense when no any test and no agent, it is OK to output 0 for both pod and mem/cpu/disk. for example:
```
Your requirement is:
    API Simple:      0
    API Script:       0 
    Browser Test:  0
    Install Agent:   N

The estimated sizing is:
    CPU:  0m
    Mem: 0MiB
    Disk:  0GB

Recommend Playback engine pods are:
    http           playback engines: 0 pod
    javascript  playback engines: 0 pod
    browser     playback engines: 0 pod
```

**Review by rongzhus — APPROVED** [review]
LGTM.

**Pull request merged**